noir-projects/noir-protocol-circuits/crates/types/src/merkle_tree/membership.nr

```diff
@@ -0,0 +1,342 @@
+}
+
+pub fn assert_check_membership<N>(leaf: Field, index: Field, sibling_path: [Field; N], root: Field) {
+    assert(check_membership(leaf, index, sibling_path, root), "membership check failed");
+}
+
+struct NonMembershipCheckErrorCodeEnum {
+    NADA: u64,
+    IS_EMPTY: u64,
+    NOT_EXISTS: u64,
+    NOT_GREATER_THAN_LOW: u64,
+    NOT_LESS_THAN_NEXT: u64,
+}
+
+global NonMembershipCheckErrorCode = NonMembershipCheckErrorCodeEnum {
+    NADA: 0,
+    IS_EMPTY: 1,
+    NOT_EXISTS: 2,
+    NOT_GREATER_THAN_LOW: 3,
+    NOT_LESS_THAN_NEXT: 4,
+};
+
+fn check_non_membership_inner<TREE_HEIGHT, LEAF_PREIMAGE>(
+    key: Field,
+    low_leaf_preimage: LEAF_PREIMAGE,
+    low_leaf_membership_witness: MembershipWitness<TREE_HEIGHT>,
+    tree_root: Field
+) -> u64 where 
+    LEAF_PREIMAGE: IndexedTreeLeafPreimage {
+    let low_key = low_leaf_preimage.get_key();
+    let next_key = low_leaf_preimage.get_next_key();
+    let is_empty_leaf = (low_key == 0) & (next_key == 0);
+
+    let low_leaf_exists = check_membership(
+        low_leaf_preimage.as_leaf(),
+        low_leaf_membership_witness.leaf_index,
+        low_leaf_membership_witness.sibling_path,
+        tree_root
+    );
+
+    if is_empty_leaf {
+        NonMembershipCheckErrorCode.IS_EMPTY
+    } else if !low_leaf_exists {
+        NonMembershipCheckErrorCode.NOT_EXISTS
+    } else if !low_key.lt(key) {
+        NonMembershipCheckErrorCode.NOT_GREATER_THAN_LOW
+    } else if !key.lt(next_key) & (next_key != 0) {
+        NonMembershipCheckErrorCode.NOT_LESS_THAN_NEXT
+    } else {
+        NonMembershipCheckErrorCode.NADA
+    }
+}
+
+pub fn check_non_membership<TREE_HEIGHT, LEAF_PREIMAGE>(
+    key: Field,
+    low_leaf_preimage: LEAF_PREIMAGE,
+    low_leaf_membership_witness: MembershipWitness<TREE_HEIGHT>,
+    tree_root: Field
+) -> bool where 
+    LEAF_PREIMAGE: IndexedTreeLeafPreimage {
+    let error = check_non_membership_inner(key, low_leaf_preimage, low_leaf_membership_witness, tree_root);
+    error == NonMembershipCheckErrorCode.NADA
+}
+
+pub fn assert_check_non_membership<TREE_HEIGHT, LEAF_PREIMAGE>(
+    key: Field,
+    low_leaf_preimage: LEAF_PREIMAGE,
+    low_leaf_membership_witness: MembershipWitness<TREE_HEIGHT>,
+    tree_root: Field
+) where 
+    LEAF_PREIMAGE: IndexedTreeLeafPreimage {
+    let error = check_non_membership_inner(key, low_leaf_preimage, low_leaf_membership_witness, tree_root);
+    if error != NonMembershipCheckErrorCode.NADA {
+        assert(
+            error != NonMembershipCheckErrorCode.IS_EMPTY, "Cannot check non membership against empty leaf"
+        );
+        assert(error != NonMembershipCheckErrorCode.NOT_EXISTS, "Low leaf does not exist");
+        assert(
+            error != NonMembershipCheckErrorCode.NOT_GREATER_THAN_LOW, "Key is not greater than the low leaf"
+        );
+        assert(
+            error != NonMembershipCheckErrorCode.NOT_LESS_THAN_NEXT, "Key is not less than the next leaf"
+        );
+    }
```

Comment:
We could add an exhaustive check here in case we add more items to the emulated enum and we forgot to add the corresponding assertion here. We could just add at the end an `assert(false, "Unknown...`

Reply:
Great idea!